From e22eaa67beb4122e8b3233b9519a82e028bee30d Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:43:57 +0800
Subject: [PATCH 4/7] GSI Android 17 compatibility:
 0004-Exfat-can-be-mounted-with-exfat-kernel-fs-driver-or-.patch

Change-Id: I35d6eb45f16d68e96bb2edd404e2433e370c6ff5
---
 fs/Exfat.cpp | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/fs/Exfat.cpp b/fs/Exfat.cpp
index 6444178..e2c4713 100644
--- a/fs/Exfat.cpp
+++ b/fs/Exfat.cpp
@@ -35,7 +35,9 @@ static const char* kFsckPath = "/system/bin/fsck.exfat";
 
 bool IsSupported() {
     return access(kMkfsPath, X_OK) == 0 && access(kFsckPath, X_OK) == 0 &&
-           IsFilesystemSupported("exfat");
+           (IsFilesystemSupported("exfat") ||
+            IsFilesystemSupported("texfat") ||
+            IsFilesystemSupported("sdfat"));
 }
 
 status_t Check(const std::string& source) {
@@ -64,13 +66,16 @@ status_t DoMount(const std::string& source, const std::string& target, int owner
     auto mountData = android::base::StringPrintf("uid=%d,gid=%d,fmask=%o,dmask=%o", ownerUid,
                                                  ownerGid, permMask, permMask);
 
-    if (mount(source.c_str(), target.c_str(), "exfat", mountFlags, mountData.c_str()) == 0) {
+    const char* fs = "exfat";
+    if (IsFilesystemSupported("sdfat")) fs = "sdfat";
+    if (IsFilesystemSupported("texfat")) fs = "texfat";
+    if (mount(source.c_str(), target.c_str(), fs, mountFlags, mountData.c_str()) == 0) {
         return 0;
     }
 
     PLOG(ERROR) << "Mount failed; attempting read-only";
     mountFlags |= MS_RDONLY;
-    if (mount(source.c_str(), target.c_str(), "exfat", mountFlags, mountData.c_str()) == 0) {
+    if (mount(source.c_str(), target.c_str(), fs, mountFlags, mountData.c_str()) == 0) {
         return 0;
     }
 
